ci/cron/src/Main.hs

```diff
@@ -113,34 +113,77 @@ to_v s = case map read $ Split.splitOn "." s of
+            putStrLn "  Checking for existing folder..."
+            old_version_exists <- exists old version
+            if old_version_exists
+            then do
+                -- Note: this checks for upload errors; this is NOT in any way
+                -- a protection against tampering at the s3 level as we get the
+                -- checksums from the s3 bucket.
+                putStrLn "  Found. Checking integrity..."
+                checksums_match <- checksums old version
+                if checksums_match
+                then do
+                    putStrLn "  Checks, reusing existing."
+                    copy (old <> "/" <> version) $ new <> "/" <> version
+                else do
+                    putStrLn "  Check failed. Rebuilding..."
+                    build version new
+            else do
+                putStrLn "  Not found. Building..."
+                build version new
+            putStrLn $ "Done " <> version <> "."
+        putStrLn $ "Copying latest (" <> latest <> ") to top-level..."
+        copy (new <> "/" <> latest <> "/*") (new <> "/")
+        putStrLn "Creating versions.json..."
+        create_versions_json versions new
+        return new
+    where
+        restore_sha io = do
+            cur_sha <- init <$> shell "git rev-parse HEAD"
+            result <- io
+            shell_ $ "git checkout " <> cur_sha
+            return result
+        download_existing_site_from_s3 path = do
+            shell_ $ "mkdir -p " <> path
+            shell_ $ "aws s3 sync s3://docs-daml-com/ " <> path
+        exists dir name = do
+            dir_exists <- Directory.doesDirectoryExist $ dir <> "/" <> name
+            dir_has_checksum_file <- Directory.doesFileExist $ dir <> "/" <> name <> "/checksum"
+            return $ dir_exists && dir_has_checksum_file
+        checksums path version = do
+            let cmd = "cd " <> path <> "/" <> version <> "; sha256sum -c checksum"
+            (code, _, _) <- shell_exit_code cmd
+            case code of
+                Exit.ExitSuccess -> return True
+                _ -> return False
+        copy from to = do
+            shell_ $ "cp -r " <> from <> " " <> to
+        build version path = do
+            shell_ $ "git checkout v" <> version
+            robustly_download_nix_packages
+            shell_ "bazel build //docs:docs"
+            shell_ $ "mkdir -p  " <> path <> "/" <> version
+            shell_ $ "tar xzf bazel-bin/docs/html.tar.gz --strip-components=1 -C" <> path <> "/" <> version
+            checksums <- shell $ "cd " <> path <> "/" <> version <> "; find . -type f -exec sha256sum {} \\;"
+            writeFile (path <> "/" <> version <> "/checksum") checksums
+        create_versions_json versions path = do
+            -- Not going through Aeson because it represents JSON objects as
+            -- unordered maps, and here order matters.
+            let versions_json = versions
```

Comment:
Order-sensitive JSON maps seem quite unintuitive to me. Can we use an array of key-value pairs instead?

Reply:
It's the js bit on the website that is order-sensitive in its JSON parsing. I'm not a big fan either, but I don't have a viable plan for changing that without breaking all the past docs.

Fun fact: we only need an ordered list of versions, so a list of strings would work here, we don't even need it to be pairs.